providers/fips/fipsprov.c

```diff
@@ -122,65 +123,80 @@ static OSSL_PARAM core_params[] =
 };
 
 /*
- * This routine is currently necessary as bn params are currently processed
- * using BN_native2bn when raw data is received. This means we need to do
- * magic to reverse the order of the bytes to match native format.
- * The array of hexdata is to get around compilers that dont like
+ * Convert a string into a bignumber.
+ * The array of hex_data is used to get around compilers that dont like
  * strings longer than 509 bytes,
  */
-static int rawnative_fromhex(const char *hex_data[],
-                             unsigned char **native, size_t *nativelen)
+static int hextobn(const char *hex_data[], BIGNUM **bn)
 {
     int ret = 0;
-    unsigned char *data = NULL;
-    BIGNUM *bn = NULL;
-    int i, slen, datalen, sz;
+    int i, slen;
     char *str = NULL;
 
+    /* Get the total length of the strings */
     for (slen = 0, i = 0; hex_data[i] != NULL; ++i)
         slen += strlen(hex_data[i]);
+
+    /* Add 1 for the string terminator */
     str = OPENSSL_zalloc(slen + 1);
     if (str == NULL)
         return 0;
+
+    /* join the strings together into 1 buffer */
     for (i = 0; hex_data[i] != NULL; ++i)
         strcat(str, hex_data[i]);
 
-    if (BN_hex2bn(&bn, str) <= 0)
-        return 0;
+    if (BN_hex2bn(bn, str) <= 0)
+        goto err;
+    ret = 1;
+err:
+    OPENSSL_free(str);
+    return ret;
+}
 
-    datalen = slen / 2;
-    data = (unsigned char *)str; /* reuse the str buffer */
+static int hextobin(const char *hex_data[], unsigned char **out, size_t *len)
+{
+    int ret = 0, sz;
+    BIGNUM *bn = NULL;
+    unsigned char *buf = NULL;
 
-    sz = BN_bn2nativepad(bn, data, datalen);
-    if (sz <= 0)
+    if (!hextobn(hex_data, &bn))
+        return 0;
+    sz = BN_num_bytes(bn);
+    buf = OPENSSL_zalloc(sz);
+    if (buf == NULL)
+        goto err;
+    if (BN_bn2binpad(bn, buf, sz) <= 0)
         goto err;
+
+    *out = buf;
+    *len = sz;
+    buf = NULL; /* Set to NULL so it is not freed */
     ret = 1;
-    *native = data;
-    *nativelen = datalen;
-    data = NULL; /* so it does not get freed */
 err:
+    OPENSSL_free(buf);
     BN_free(bn);
-    OPENSSL_free(data);
     return ret;
 }
 
-/* TODO(3.0): To be removed */
-static int dummy_evp_call(void *provctx)
+#ifndef OPENSSL_NO_DSA
```

Comment:
A wider question: should the FIPS provider include conditional checks for `OPENSSL_NO_DSA` and other macros?

They are beneficial for 3.0 (and likely validations beyond that).  They complicate the source code.  They could be useful to third parties.

Reply:
Yes. If someone compiles with "no-dsa" then that is what they should get - including in the fips module.